### ISSUE TITLE
Inability to reproduce numbers from article

### ISSUE THREAD

**Issue by Shnurre** [clab/stack-lstm-ner#7]
Hello,

Thank you for your most interesting article.

I have successfully built your code and tried to implement it on the CoNLL-2003 dataset. However, the numbers I was able to receive are significantly lower than the ones reported in the article. I get around 0.86 on dev-set, 0.828 on test-set using built-in comparator and 0.795 using conlleval.
I run the code exactly as it is described in readme.

 The only difference is that I have sskip.100.vector in subfolder sskip, so I have `-w sskip/sskip.100.vectors --pretrained_dim 100` `instead of -w sskip.100.vectors --pretrained_dim 100`, but I don't think it could make a difference. I also run the code directly and not in the screen mode as the readme suggests.

However, it is possible that embeddings were not used in my run(since the number I received resembles the one reported in the article without the embeddings). During my run I see `Loading from sskip/sskip.100.vectors with100 dimensions` followed by `REDUCE(ORG) :: 0`.  Is there a way to verify that pretrained embeddings were indeed loaded successfully? Judging by the code there shouldn't be any additional messages that pretained embeddings were loaded successfully.

The other reason of low numbers may be that I have trained the mode for too long or not long enough. I couldn't find any information on how many epochs should pass before model is ready so I trained it for approximately 20 hours (the last training epoch was 94.275, update 14129). If it is the wrong number of epochs could you please specify what is the correct one?

Sorry to bother you it just your article looked really interesting and I wanted to reproduce the results)

**Comment by miguelballesteros** [comment]
Hi @Shnurre 

what is the evaluation script that you are using? The system will output a number but this is not necessarily the same as the one that conlleval will output.

**Have you used conlleval for evaluating the final output?**

You should train it until there are no changes in the model file for 6 hours or so.

**Comment by miguelballesteros** [comment]
Can you send me your exact command?

**Comment by Shnurre** [comment]
Hello, @miguelballesteros. Thank you for your quick response.

Sure, here is my command `ismurov@smurov_unix:~/stack-lstm-ner-emb/stack-lstm-ner/build/ner-system$ ./lstm-parse -T conll/train.parser -d conll/dev.parser --hidden_dim 100 --lstm_input_dim 100 -w sskip/sskip.100.vectors --pretrained_dim 100 --rel_dim 20 --action_dim 20 --input_dim 100 -t -S -D 0.3 > logNERYesCharNoPosYesEmbeddingsD0.3.txt`.

Yes, I have used conlleval for evaluating the final output. The result is f-measure 0.795.

0.828 is the f-measure shown when running `./lstm-parse -T conll/train.parser -d conll/test.parser --hidden_dim 100 --lstm_input_dim 100 -w sskip/sskip.100.vectors --pretrained_dim 100 --rel_dim 20 --action_dim 20 --input_dim 100 -m latest_model -S > output.txt`. 

The results on dev-set were between 0.855 and 0.865 for several hours but I suppose I could wait longer.

**Comment by miguelballesteros** [comment]
Thanks @Shnurre 

when I saw your message I downloaded the code again, and I trained a model for just 1 hour (for the paper we trained the models for 24 hours or so). I just got this result in the test set:

processed 46666 tokens with 5648 phrases; found: 5581 phrases; correct: 4912.
accuracy:  97.39%; precision:  88.01%; recall:  86.97%; FB1:  87.49
              LOC: precision:  90.95%; recall:  87.95%; FB1:  89.42  1613
             MISC: precision:  78.14%; recall:  70.80%; FB1:  74.29  636
              ORG: precision:  84.42%; recall:  85.13%; FB1:  84.77  1675
              PER: precision:  92.58%; recall:  94.87%; FB1:  93.71  1657

which looks like a lot more than what you are getting. **Remember that this is trained for just 1 hour.** so it should go up or close to 90 in some hours

These are my commands:

./lstm-parse -T conll2003/train.parser -d conll2003/dev.parser --hidden_dim 100 --lstm_input_dim 100 -w sskip.100.vectors --pretrained_dim 100 --rel_dim 20 --action_dim 20 --input_dim 100 -D 0.2 -t -S > log.txt &

./lstm-parse -T conll2003/train.parser -d conll2003/test.parser --hidden_dim 100 --lstm_input_dim 100 -w sskip.100.vectors --pretrained_dim 100 --rel_dim 20 --action_dim 20 --input_dim 100 -D 0.2 -m latest_model -S > out.txt

python attach_prediction.py -p out.txt -t conll2003/test -o os.txt
./conlleval < os.txt 

**So, I think that you are probably not using the correct embeddings. Where did you get yours? Are you sure that the path is correct?**

**Comment by Shnurre** [comment]
Thank you, @miguelballesteros . 
Yes, the numbers I get look like the ones in the article without embeddings, which I mentioned in my first post. 
I don't know what could be wrong with embeddings though. I downloaded them using link in readme (https://drive.google.com/file/d/0B8nESzOdPhLsdWF2S1Ayb1RkTXc/view). It has 243004 lines and looks what you expect embedding file to look like
I am sure the path is correct but I probably should copy the file to ner-system directory and use the exact command you are using. I'll run it now and report my numbers after couple hours of training

**Comment by miguelballesteros** [comment]
My embeddings file has 243004 lines yes. So it looks like you have the correct file. 
You can use any path as soon as the path is correct, you can also do a quick test as I did... you should get something similar to what I posted in the previous message

**Comment by miguelballesteros** [comment]
after 6 hours I've got 89.60, you should get the same

**Comment by Shnurre** [comment]
This time I used the exact commands you did and trained for approximately  17 hours. 
It appears that emeddings are still not used. I got the following results:

On test:
processed 46666 tokens with 5648 phrases; found: 5722 phrases; correct: 4592.
accuracy:  96.19%; precision:  80.25%; recall:  81.30%; FB1:  80.77
LOC: precision:  85.71%; recall:  85.91%; FB1:  85.81  1672
MISC: precision:  73.73%; recall:  70.37%; FB1:  72.01  670
ORG: precision:  74.44%; recall:  78.57%; FB1:  76.45  1753
PER: precision:  83.59%; recall:  84.11%; FB1:  83.85  1627

On dev:
processed 51578 tokens with 5942 phrases; found: 5924 phrases; correct: 5098.
accuracy:  97.47%; precision:  86.06%; recall:  85.80%; FB1:  85.93
LOC: precision:  89.25%; recall:  91.29%; FB1:  90.26  1879
MISC: precision:  87.79%; recall:  81.13%; FB1:  84.33  852
ORG: precision:  80.06%; recall:  79.94%; FB1:  80.00  1339
PER: precision:  86.35%; recall:  86.92%; FB1:  86.63  1854

On train:
processed 204567 tokens with 23499 phrases; found: 23440 phrases; correct: 23345.
accuracy:  99.93%; precision:  99.59%; recall:  99.34%; FB1:  99.47
LOC: precision:  99.69%; recall:  99.59%; FB1:  99.64  7133
MISC: precision:  98.97%; recall:  97.82%; FB1:  98.39  3398
ORG: precision:  99.63%; recall:  99.32%; FB1:  99.48  6301
PER: precision:  99.77%; recall:  99.89%; FB1:  99.83  6608

I will try deleting and rebuiling everything and if it won't help try to build it on another machine.

**Comment by miguelballesteros** [comment]
I used the exact same code as you (if you haven't changed anything), and my result was (after 6 hours):

processed 46666 tokens with 5648 phrases; found: 5767 phrases; correct: 5114.
accuracy:  97.68%; precision:  88.68%; recall:  90.55%; FB1:  89.60
              LOC: precision:  92.94%; recall:  91.61%; FB1:  92.27  1644
             MISC: precision:  75.17%; recall:  78.92%; FB1:  77.00  737
              ORG: precision:  85.24%; recall:  89.70%; FB1:  87.42  1748
              PER: precision:  94.14%; recall:  95.36%; FB1:  94.75  1638

If you have doubts about whether you are using pretrained embeddings or not, you can print what is going here: https://github.com/clab/stack-lstm-ner/blob/master/ner-system/lstm-parse.cc#L195 or here https://github.com/clab/stack-lstm-ner/blob/master/ner-system/lstm-parse.cc#L500

The machine should not have any effect at all.
